### ISSUE TITLE
Los tiempos mostrados varían según el usuario

### ISSUE THREAD

**Issue by danmarsden** [catalyst/moodle-block_dedication#49]
**[Original report](https://bitbucket.org/ciceidev/moodle_block_dedication/issue/49) by Anonymous.**

----------------------------------------

Buenos días.

Acabamos de darnos cuenta que el tiempo de dedicación que le muestra a un alumno no es el mismo que se le muestra a un usuario gestor, u otro rol con permisos.

En concreto, para este alumno, veo que se muestra una sesión con dos ips. ¿Es posible que el problema venga por ahí?

Gracias.

Saludos.

**Comment by danmarsden** [comment]
needs more info to reproduce - closing for now but we can re-open later if needed.

**Comment by rmartinez-c2** [comment]
What type of info do you need?

I've attached some images for more explanation
![01  Tiempo estudiante](https://user-images.githubusercontent.com/3283956/202651255-41d54d9f-3bca-4cea-b189-0738e4bbf18b.png)
![02  Tiempos estudiantes visto por gestor](https://user-images.githubusercontent.com/3283956/202651258-164edea4-82f0-4772-b9ad-e2eb16f45c6e.png)

The first one is the dedication time the user can view and the second one is the dedication time of the entire course.

Next time, could I write in spanish??

Thanks in advance.
Regards.

**Comment by rmartinez-c2** [comment]
One last screenshot of the dedication time student in detail

![03  Tiempo estudiante detallado](https://user-images.githubusercontent.com/3283956/202652218-73140818-b145-4909-9660-0513b56b7b7d.png)

As you can see, there are two rows with two ips. Could that be the source of the problem?

**Comment by rmartinez-c2** [comment]
Moodle 3.7.1 (Build: 20190708)
PHP Version 7.2.20
block_dedication: Moodle 3.x+ (2020041000)

**Comment by danmarsden** [comment]
Hi @rmartinezgc thanks for adding the screenshots! - we're currently in the process of rewriting a lot of this code to make it perform better (see the readme file on the 4.0 branch for more info on some of the plans)

I expect the work we're doing will fix this as we are making it a lot more consistent how it's generating the values.  I don't expect it to be ready for testing for at least a couple more weeks, but we also have a number of other plans to improve the way the times are being calculated as well as adding supporf for Moodle's report-builder apis making it easier to report across courses.

Once we've finished the updates it would be great if you were able to test our new plugin and see if it's giving you more consistent reporting values.

I won't hold it against you if you post in spanish (I'll just be using Google Translate to translate it) but I'll only be posting in English here myself so you might need to do the same (use google translate to translate my text)


thanks again!

**Comment by rmartinez-c2** [comment]
Will the new version be compatible with our Moodle?

On the other hand, the problem we have is not that the sum of dedication time is incorrect but that it seems that in the student view it takes into account more actions than when reviewed by a manager. Could you help me with the current version of the plugin, or what part of the code I could check or what dedication time is the right one?

Thanks in advance.
Regards.

**Comment by danmarsden** [comment]
The version we are currently working on will only support Moodle 4.0 and higher as it uses some of the reporting features that are available in Moodle 4.0.

I suspect the issue you are having is due to the way the reporting page and block use different settings - this is one of the things we're looking at tidying up - try viewing the block edit settings page to see if the settings there match up with the ones you are using on the report page. Keep in mind - we've only just taken over maintenance of this plugin in the last few days, and we're heads-down working on the "new" version as part of funding from one of our clients. If that still doesn't help, you could try asking for help in the community forums to see if anyone else using this plugin has seen the issue or solved it.

**Comment by rmartinez-c2** [comment]
I've just found the problem. The calculations are done well, the problem is that the course has the limit set to 120 minutes and, by default, the manager form calculates it with a limit of 60 minutes. Ideally, the form should indicate the value configured in the course by default to avoid misunderstandings like this. Reviewing the code, I find that the select limit defaults to BLOCK_DEDICATION_DEFAULT_SESSION_LIMIT. How could I assign the value set in the course?

The line in question, that I think should be changed, is line 73 of the dedication/dedication.php file:

`$limit = optional_param('limit', BLOCK_DEDICATION_DEFAULT_SESSION_LIMIT, PARAM_INT);`

Thank you in advance.

Regards.

**Comment by rmartinez-c2** [comment]
Solved with the following code:

```
// CoDeMan(2022-11-21): Por defecto, asignamos el límite al valor configurado en el curso.
// En caso de no existir, asignamos el valor por defector general BLOCK_DEDICATION_DEFAULT_SESSION_LIMIT
//$limit = optional_param('limit', BLOCK_DEDICATION_DEFAULT_SESSION_LIMIT, PARAM_INT);

$coursecontext = context_course::instance($course->id);
$blockrecord = $DB->get_record(
	'block_instances', 
	array('blockname' => 'dedication','parentcontextid' => $coursecontext->id)
	, '*' , MUST_EXIST
);
$blockinstance = block_instance('dedication', $blockrecord);
$ilimit = $blockinstance->config->limit ?? BLOCK_DEDICATION_DEFAULT_SESSION_LIMIT;
$limit = optional_param('limit', $ilimit, PARAM_INT);
// ---
```
Regards.

**Comment by danmarsden** [comment]
nice work! - in the new version of the plugin (for 4.0 and higher) we've removed the 2 settings and combined them into one (at the site-level only) to help prevent the confusion you're seeing and we've also added some text to display on the full report page to make it transparent to the teacher viewing the report what the session limit is set to as well.
here's an example of what that currently looks like:

![Screenshot from 2022-11-22 19-34-48](https://user-images.githubusercontent.com/362798/203242014-122fbaaf-2d34-40b1-9569-6e9a01880f78.png)
The bottom item is the "session limit" being defined as 3 minutes.

What I think might also be confusing you is that the session limit is not defined in minutes, but seconds and it excludes all sessions lower than the session limit from being included in the calculations - so set to 120 it means 120 seconds, (not minutes).

We don't currently have plans to make this same change (combining the 2 settings into one) on the older branch - only in the Moodle 4.0 and higher supported branch at this stage.

thanks!
